File: go/perfect-numbers/perfect_numbers.go
package perfect

import "errors"

var ErrOnlyPositive = errors.New("no!")

type Classification int

const (
	ClassificationDeficient Classification = iota
	ClassificationPerfect
	ClassificationAbundant
)

func Classify(n int64) (Classification, error) {
	if n <= 0 {
		return 0, ErrOnlyPositive
	}

	sum := int64(0)

	for i := int64(1); i < n; i++ {
		if n%i == 0 {
			sum += i
		}
	}

	if sum == n {
		return ClassificationPerfect, nil
	} else if sum < n {
		return ClassificationDeficient, nil
	} else {
		return ClassificationAbundant, nil
	}

}

File: zig/perfect-numbers/perfect_numbers.zig
const std = @import("std");
pub const Classification = enum {
    deficient,
    perfect,
    abundant,
};
/// Asserts that `n` is nonzero.
pub fn classify(n: u64) Classification {
    std.debug.assert(n > 0);
    var sum: u64 = 0;
    for (1..n) |i| {
        if (@mod(n, i) == 0)
            sum += i;
    }
    return if (sum == n) .perfect else if (sum > n) .abundant else .deficient;
}
